### PR TITLE
filebrowser: init at 2.23.0

### LINKED ISSUE #122339: Filebrowser (https://github.com/filebrowser/filebrowser)
**Project description**
Lightweight nextcloud/seafile/syncthing/ftp alternative to easily share files with mutiple users via HTTP

**Metadata**

* homepage URL: https://github.com/filebrowser/filebrowser
* source URL: https://github.com/filebrowser/filebrowser
* license: Apache License
* platforms: any

### PR DESCRIPTION
**Pull request by nielsegberts**
###### Description of changes

Filebrowser is a web application for managing files and directories. See https://filebrowser.org/

Lightweight nextcloud/seafile/syncthing/ftp alternative to share files via HTTP.

A new service might be useful later, but this pull request just introduces the package.

Fixes https://github.com/NixOS/nixpkgs/issues/122339

This is my first package contribution to nixpkgs.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/filebrowser/default.nix [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  frontend = buildNpmPackage rec {
+    pname = "filebrowser-frontend";
+    version = "2.23.0";
+
+    src = fetchFromGitHub {
+        owner = "filebrowser";
+        repo = "filebrowser";
+        rev = "v${version}";
+        sha256 = "sha256-xhBIJcEtxDdMXSgQtLAV0UWzPtrvKEil0WV76K5ycBc=";
+    } + "/frontend";
```
Comment (RaitoBezarius) [inline pkgs/applications/networking/filebrowser/default.nix:13]: Please use `sourceRoot = "source/frontend";` rather

File: pkgs/applications/networking/filebrowser/default.nix [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    npmDepsHash = "sha256-acNIMKHc4q7eiFLPBtKZBNweEsrt+//0VR6dqwXHTvA=";
+
+    NODE_OPTIONS = "--openssl-legacy-provider";
```
Comment (RaitoBezarius) [inline pkgs/applications/networking/filebrowser/default.nix:17]: Hm, why is it needed? We have EOLed Node.js 16 which I believe is the last affected by this issue now.

File: pkgs/applications/networking/filebrowser/default.nix [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    installPhase = ''
+        runHook preInstall
+
+        mkdir -p $out/dist
+        mv dist $out/
```
Comment (RaitoBezarius) [inline pkgs/applications/networking/filebrowser/default.nix:23]: Maybe keep it simple: `mv dist $out` and don't bother with creating `dist`

File: pkgs/applications/networking/filebrowser/default.nix [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        rev = "v${version}";
+        sha256 = "sha256-xhBIJcEtxDdMXSgQtLAV0UWzPtrvKEil0WV76K5ycBc=";
+    };
+
+    vendorHash = "sha256-MR0ju2Nomb3j78Z+1YcJY+jPd40MZpuOTuQJM94AM8A=";
+
+    excludedPackages = [ "tools" ];
+
+    preBuild = ''
+      cp -r ${frontend}/dist/* frontend/dist
```
Comment (RaitoBezarius) [inline pkgs/applications/networking/filebrowser/default.nix:45]: Maybe do a symlink?

File: pkgs/applications/networking/filebrowser/default.nix [hunk 5]
```diff
@@ -0,0 +1,55 @@
+{ buildGoModule, buildNpmPackage, fetchFromGitHub, lib }:
+
+let
+  frontend = buildNpmPackage rec {
```
Comment (RaitoBezarius) [inline pkgs/applications/networking/filebrowser/default.nix:4]: This derivation should be part of the internal one as a `passthru` for example so you can do `filebrowser.frontend` or `filebrowser.ui`, look nixpkgs for existing examples.

File: pkgs/applications/networking/filebrowser/default.nix (hunk 4 above)
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix:45]: Doesn't work unfortunately, the files in dist are picked up by go's embed, and it doesn't work with symlinks

https://github.com/golang/go/issues/44507

File: pkgs/applications/networking/filebrowser/default.nix (hunk 2 above)
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix:17]: it's using Node.js v18.16.0, and I don't know why, I only know my build fails without it:

```
    at FSReqCallback.oncomplete (node:fs:198:23) {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}

Node.js v18.16.0

ERROR: `npm build` failed
```

File: pkgs/applications/networking/filebrowser/default.nix (hunk 1 above)
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix:13]: Thanks!

File: pkgs/applications/networking/filebrowser/default.nix (hunk 4 above)
Comment (RaitoBezarius) [reply pkgs/applications/networking/filebrowser/default.nix:45]: Go is tiring. :<

File: pkgs/applications/networking/filebrowser/default.nix (hunk 2 above)
Comment (RaitoBezarius) [reply pkgs/applications/networking/filebrowser/default.nix:17]: can you bump it to node 20?

File: pkgs/applications/networking/filebrowser/default.nix (hunk 3 above)
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix:23]: Done

File: pkgs/applications/networking/filebrowser/default.nix (hunk 5 above)
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix:4]: Added a passthru, please take a look

File: pkgs/applications/networking/filebrowser/default.nix (hunk 2 above)
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix:19]: I bumped it with:

```
  frontend = buildNpmPackage.override { nodejs = nodejs_20; } rec {
    pname = "filebrowser-frontend";
```

But the error remains the same:

```
    at FSReqCallback.oncomplete (node:fs:188:23) {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}

Node.js v20.2.0

ERROR: `npm build` failed
```

File: maintainers/maintainer-list.nix [hunk 6]
```diff
@@ -11393,6 +11393,12 @@
       fingerprint = "E576 BFB2 CF6E B13D F571  33B9 E315 A758 4613 1564";
     }];
   };
+  nielsegberts = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:11396]: Please put this into a separate commit with the message ``maintainers: add NAME`` and sort it alphabetically.

File: pkgs/applications/networking/filebrowser/default.nix [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        frontend = frontend;
+    };
+
+    meta = with lib; {
+        description = "Filebrowser is a web application for managing files and directories";
+        homepage = "https://filebrowser.org";
+        license = licenses.asl20;
+        maintainers = with maintainers; [ nielsegberts ];
+        platforms = platforms.unix;
+  };
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/filebrowser/default.nix:60]:
```suggestion
    passthru = {
      inherit frontend;
    };

    meta = with lib; {
      description = "Filebrowser is a web application for managing files and directories";
      homepage = "https://filebrowser.org";
      license = licenses.asl20;
      maintainers = with maintainers; [ nielsegberts ];
  };
```
platforms unix is the default

File: pkgs/applications/networking/filebrowser/default.nix [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        sha256 = "sha256-xhBIJcEtxDdMXSgQtLAV0UWzPtrvKEil0WV76K5ycBc=";
+    };
+
+    vendorHash = "sha256-MR0ju2Nomb3j78Z+1YcJY+jPd40MZpuOTuQJM94AM8A=";
+
+    excludedPackages = [ "tools" ];
+
+    preBuild = ''
+      cp -r ${frontend}/dist/* frontend/dist
+    '';
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/filebrowser/default.nix:48]:
```suggestion
buildGoModule rec {
  pname = "filebrowser";
  version = "2.23.0";

  src = fetchFromGitHub {
    owner = "filebrowser";
    repo = "filebrowser";
    rev = "v${version}";
    sha256 = "sha256-xhBIJcEtxDdMXSgQtLAV0UWzPtrvKEil0WV76K5ycBc=";
  };

  vendorHash = "sha256-MR0ju2Nomb3j78Z+1YcJY+jPd40MZpuOTuQJM94AM8A=";

  excludedPackages = [ "tools" ];

  preBuild = ''
    cp -r ${frontend}/dist/* frontend/dist
  '';
```
we use 2 spaces for nix files

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix (hunk 6 above)
Comment (nielsegberts) [reply maintainers/maintainer-list.nix:11396]: Done!

File: pkgs/applications/networking/filebrowser/default.nix (hunk 7 above)
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix]: Removed

File: pkgs/applications/networking/filebrowser/default.nix (hunk 8 above)
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix]: Changed to two spaces

File: pkgs/applications/networking/filebrowser/default.nix [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+let
+  frontend = buildNpmPackage rec {
+    pname = "filebrowser-frontend";
+    version = "2.23.0";
+
+    src = fetchFromGitHub {
+      owner = "filebrowser";
+      repo = "filebrowser";
+      rev = "v${version}";
+      sha256 = "sha256-xhBIJcEtxDdMXSgQtLAV0UWzPtrvKEil0WV76K5ycBc=";
```
Comment (drupol) [inline pkgs/applications/networking/filebrowser/default.nix:12]:
```suggestion
      hash = "sha256-xhBIJcEtxDdMXSgQtLAV0UWzPtrvKEil0WV76K5ycBc=";
```

**Review by drupol — APPROVED** [review]

**Review by drupol — CHANGES_REQUESTED** [review]
Sorry I hit the wrong select box in my previous review where I added a few comments.

File: pkgs/applications/networking/filebrowser/default.nix (hunk 9 above)
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix]: Thanks for the review, done

File: pkgs/applications/networking/filebrowser/default.nix [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+in
+buildGoModule rec {
+  pname = "filebrowser";
+  version = "2.23.0";
+
+  src = fetchFromGitHub {
+    owner = "filebrowser";
+    repo = "filebrowser";
+    rev = "v${version}";
+    sha256 = "sha256-xhBIJcEtxDdMXSgQtLAV0UWzPtrvKEil0WV76K5ycBc=";
```
[the remark this answers is not in the document]
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix]: Done

File: pkgs/applications/networking/filebrowser/default.nix [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    rev = "v${version}";
+    hash = "sha256-xhBIJcEtxDdMXSgQtLAV0UWzPtrvKEil0WV76K5ycBc=";
+  };
+
+  vendorHash = "sha256-MR0ju2Nomb3j78Z+1YcJY+jPd40MZpuOTuQJM94AM8A=";
+
+  excludedPackages = [ "tools" ];
+
+  preBuild = ''
+    cp -r ${frontend}/dist/* frontend/dist
```
Comment (drupol) [inline pkgs/applications/networking/filebrowser/default.nix:47]:
```suggestion
    cp -r ${frontend}/dist frontend/
```

**Review by drupol — CHANGES_REQUESTED** [review]
I just add a small remark without being able to test it right now. Could you test it on your side with that change?

File: pkgs/applications/networking/filebrowser/default.nix (hunk 11 above)
Comment (nielsegberts) [reply pkgs/applications/networking/filebrowser/default.nix]: Yep that works, thanks

**Review by drupol — APPROVED** [review]
LGTM, please ping me when eval is done so I can merge it.

**Pull request merged**